Feature month segregation for project report list

maciejSamerdak opened the pull request:
**project-report-list** view now accepts a month and year parameters to separate reports by each month. A navigation bar is also introduced, containing next and previous button to navigate between adjacent months, as well as form allowing to select desired month and year to jump to.

Related #137 

**Do not delete this branch until #139 and #256 are merged**

rwrzesien commented:
@maciejSamerdak Same here, could add few words explaining what is it doing?

If I understood correctly, this should be a view for going through months back and forth and then it load report for that month?

If yes, then going from 01.2000 to 01.2099 would require 99 * 12 requests? Or can I go directly to the requested date?

maciejSamerdak commented:
I updated the description. Navigation bar includes form allowing to go directly to specified month.

maciejSamerdak commented on employees/common/constants.py:
```diff
@@ -9,3 +9,9 @@ class ReportModelConstants(Enum):
+MONTH_NAVIGATION_FORM_MAX_MONTH_VALUE = 12
+MONTH_NAVIGATION_FORM_MIN_MONTH_VALUE = 1
+MONTH_NAVIGATION_FORM_MAX_YEAR_VALUE = 2099
+MONTH_NAVIGATION_FORM_MIN_YEAR_VALUE = 1990
```
I figured this could be the year CodePoets was founded, but I forgot to ask about it and shoot me dead guys, I couldn't remember for the life of me, which anniversary we had back in autumn (5th, I checked now...). Unless it's more sturdy to start from 2000?

rwrzesien commented:
@maciejSamerdak 
1. Please rebase this branch on `master` after merging base branch, there are currently too many things in this pull request that are not related so its hard to review.
2. I like the general idea, but how about simplifying the UX part and using configured datepicker, and using its callback (https://api.jqueryui.com/datepicker/#option-onChangeMonthYear) to load the requested month data using AJAX call and replace current data? I think this could result in having less code in the end. But I am not sure until you apply point 1.

rwrzesien commented:
@maciejSamerdak Please apply https://github.com/Code-Poets/sheetstorm/pull/138#issuecomment-491627651 , lets start with making rebase to see only changes from this pull request.

maciejSamerdak commented:
I've ran into multiple problems on frontend while trying to implement form-based month switch on UX side, as suggested by @rwrzesien. I've spent entire day and sadly, I couldn't come to any sensible solution, so I had to rollback to previous POST related solution.

I've refactored entire feature a bit, moving the entire functionality from an **inclusion tag** to a **view mixin**. The code is basically the same, both the feature and it's test, I just moved it elsewhere. Still, I believe a complete re-review will be necessary. **The cons:** rather than render the navbar template from inclusion tag, we have to remember to include it manually in target template. **The pros:** the code might be considered cleaner, since all we have to do now for a Django-based view to include the navbar, is to simply join the mixin

dybi commented on employees/views.py:
```diff
@@ -49,6 +59,70 @@ def perform_create(self, serializer: ReportSerializer) -> None:
+        return context_data
+
+    def redirect_to_another_month(self, request: HttpRequest) -> HttpResponseRedirectBase:
+        post_data = request.POST.copy()
```
If I get this right, this function redirect to another month, if the form is valid. I would suggest changing the logic a bit:
```python
post_data = request.POST
        form = MonthSwitchForm(data=post_data)
        if not form.is_valid():
            return redirect(request.path)

        new_date = datetime.datetime.strptime(post_data["date"], "%m-%Y")
        return redirect(
            reverse(
                self.url_name,
                kwargs={"pk": self.kwargs["pk"], "year": new_date.year, "month": new_date.month},
            )
        )
```

it seems more readable to me and does nt copy anything.

maciejSamerdak replied:
The `form.is_valid()` will fail every time, because date in POST is passed as a string and the `MonthSwitchForm` doesn't handle that and needs a date input instead. This is why I switched the date in `request.POST` before passing it to form, but I had to clone it first to have it mutable.

dybi commented on managers/views.py:
```diff
@@ -69,6 +70,12 @@ class ProjectDetailView(DetailView):
 
+    def get_context_data(self, **kwargs: Any) -> dict:
+        context = super().get_context_data(**kwargs)
+        context["year"] = timezone.now().year
```
maybe a silly question, but why do we use `timezone.now()` here when in (almost?) all other cases we use `datetime.now()`?

maciejSamerdak replied:
Generally, I treat them interchangeably, since timezones are not a big deal for the most part. There was one instance when `timezone` had to be used over `datetime` in unit tests, because of issues with `freezegun` when comparing timestamps. Majority of files I've been working on were already using datetime, so I just stuck to it. Here, there was no such module before, so I picked `timezone`. I guess I just consider it more reliable. I also like the simpler syntax, instead of awkward double **datetime**.

Yeah, I guess I could just import `date` and `datetime` separately and refactor the existing code to avoid it...
Or perhaps, if we were to change the use of `datetime` in code, we should drop it altogether in favour of Django's `timezone`?

dybi replied:
maybe we could refactor out code to use `django`'s `timezone`, but I think it should be another issues.
@rwrzesien what do you think?

rwrzesien replied:
@dybi Yes, there was an issue for that I think to change all usages to `timezone`. Basically it does the same as datetime, but takes into consideration Django timezone settings (defined in settings file). So if at some point of the project we decide to use timezone(s) or change current timezone it is already done everywhere already. So I would say its just the matter of good Django practice.

maciejSamerdak commented on employees/tests/test_unit_custom_mixins.py:
```diff
@@ -0,0 +1,132 @@
+        }
+        for key in expected_output.keys():  # pylint: disable=consider-iterating-dictionary
+            if key == "month_form":
+                self.assertTrue(expected_output[key].__eq__(context.dicts[1][key]))
```
There's an issue with the `MonthSwitchForm` object. It's the only object that cannot be compared with `==` operator, `self.assertEquals` fails on this one.

dybi replied:
the funny thing is that `==` is basically `obj.__eq__(some_other_obj)` ;) I checked it with debbuger: `expected_output[key].__eq__(context.dicts[1][key])` returns `NotImplemented` which's value is different than `False`.

To sum up: blease implement `__eq__` method in `MonthSwitchForm` and use normal `assertEquals` in the test ;)

maciejSamerdak replied:
Done

maciejSamerdak commented on employees/tests/test_unit_custom_mixins.py:
```diff
@@ -0,0 +1,132 @@
+            "title_date": "03/19",
+        }
+        for key in expected_output.keys():  # pylint: disable=consider-iterating-dictionary
+            if key == "month_form":
```
`MonthNavigationText` class under presented under **navigation_text** cannot be compared using '__eq__', it causes an error in test (I guess it's because it's not an instance?).
`MonthSwitchForm` is the only object in the whole package that causes an issue on `self.assertEqual()`, therefore there's this differentiation.

dybi replied:
please see above

maciejSamerdak replied:
Done

maciejSamerdak commented on employees/common/constants.py:
```diff
@@ -1,6 +1,11 @@
+MONTH_NAVIGATION_FORM_MAX_MONTH_VALUE = 12
+MONTH_NAVIGATION_FORM_MIN_MONTH_VALUE = 1
+MONTH_NAVIGATION_FORM_MAX_YEAR_VALUE = 2099
+MONTH_NAVIGATION_FORM_MIN_YEAR_VALUE = 2000
```
Well, if there are no plans for filling up the database with previous reports for some _legacy purposes_...

maciejSamerdak replied:
Done

maciejSamerdak commented on employees/views.py:
```diff
@@ -49,6 +59,70 @@ def perform_create(self, serializer: ReportSerializer) -> None:
+    def _get_month_navigator_params(self) -> dict:
+        disable_next_button = False
+        disable_previous_button = False
+        year = int(self.kwargs["year"])
```
`kwargs` are attached to the view and they will be used once those methods are called from the view which use the mixin. The `ContextMixin` itself does not contain it sadly.

Should I override it's `__init__` to include the `kwargs` attribute or add it the way I did the `url_name`?

dybi replied:
I would prefer the latter.
@rwrzesien What would the most `Django` way to do it? ;)

maciejSamerdak replied:
Did it the latter way

rwrzesien replied:
> @rwrzesien What would be the most Django way to do it? ;)

Both ways are fine I think, but as we don't do anything at the `__init__` at the moment I would also prefer the way which you have chosen.

maciejSamerdak commented on employees/tests/test_unit_custom_mixins.py:
```diff
@@ -0,0 +1,132 @@
+from employees.views import MonthNavigationMixin
+
+
+class MonthNavigationMixinCustomCustomMethodsTests(TestCase):
```
Thanks for pointing that out! The whole `mixin` variable was a rushed fix prior to push, after I changed these methods from static to regular. I must've been too tired after fiddling with this feature for whole day and missed this opportunity. 😫

maciejSamerdak replied:
Done

maciejSamerdak commented on employees/forms.py:
```diff
@@ -22,3 +29,30 @@ class Meta:
+    )
+
+    def __init__(  # pylint: disable=keyword-arg-before-vararg
+        self, initial_date: datetime.date = None, *args: Any, **kwargs: Any
```
Was this solution meant to silence pylint? Because it still argues over it. ☹️ 
Mypy doesn't mind either way...

dybi replied:
well, it's weird - it looks like it doesn't complain on mine ;)
As to the root of the problem - `initial_date` is of `Optional[datetime.date]`, because it can be `None`: https://mypy.readthedocs.io/en/latest/kinds_of_types.html#optional-types-and-the-none-type

maciejSamerdak commented on employees/tests/test_unit_custom_mixins.py:
```diff
@@ -0,0 +1,132 @@
+from employees.views import MonthNavigationMixin
+
+
+class MonthNavigationMixinCustomCustomMethodsTests(TestCase):
```
Done

rwrzesien commented:
@maciejSamerdak 
> I've ran into multiple problems on frontend while trying to implement form-based month switch on UX side, as suggested by @rwrzesien.

I rather meant using JS datepicker + callback for loading other months data via AJAX call, but you did a lot of work here so lets keep current approach.
